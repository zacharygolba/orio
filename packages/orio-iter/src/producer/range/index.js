// @flow

export { default as CharRange } from './char'
export { default as NumberRange } from './number'
