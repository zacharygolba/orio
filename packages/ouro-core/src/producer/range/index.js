// @flow

export { default as Chars } from './chars'
export { default as Numbers } from './numbers'
