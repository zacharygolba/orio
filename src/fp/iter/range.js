// @flow

export default function* range(start?: number = 0, end?: number = Infinity): Iterator<number> {
  for (let next = start; next < end; next += 1) {
    yield next
  }
}
