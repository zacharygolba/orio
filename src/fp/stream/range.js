// @flow

export default async function* range(
  start?: number = 0,
  end?: number = Infinity,
): AsyncIterator<number> {
  for (let next = start; next < end; next += 1) {
    yield next
  }
}
