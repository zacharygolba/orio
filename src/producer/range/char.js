// @flow

import * as result from '../../result'
import type { Producer } from '../types'

import NumberProducer from './number'

export const MIN_CHAR: string = String.fromCodePoint(0)
export const MAX_CHAR: string = String.fromCodePoint(0x10ffff)

export default class CharProducer implements Producer<*> {
  source: NumberProducer
  size: number
  /*:: @@iterator: () => Iterator<string> */

  constructor(start?: string = MIN_CHAR, end?: string = MAX_CHAR) {
    this.source = new NumberProducer(start.codePointAt(0), end.codePointAt(0))
    this.size = this.source.size
  }

  // $FlowFixMe
  [Symbol.iterator](): Iterator<string> {
    return this
  }

  next(): IteratorResult<string, void> {
    const next = this.source.next()

    if (next.done) {
      return next
    }

    return result.next(String.fromCodePoint(next.value))
  }

  sizeHint(): number {
    return this.size
  }
}
