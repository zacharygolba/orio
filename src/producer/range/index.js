// @flow

export { default as CharProducer } from './char'
export { default as NumberProducer } from './number'
