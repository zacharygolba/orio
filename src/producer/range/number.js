// @flow

import * as result from '../../result'
import ProducerBase from '../base'

export default class NumberProducer extends ProducerBase<number> {
  done: boolean
  end: number
  size: number
  start: number
  state: number
  step: number

  constructor(s?: number = 0, e?: number = Infinity) {
    super()

    const start = Number(s)
    const end = Number(e)

    this.size = Math.abs(start - end) + 1

    if (Number.isNaN(start) || Number.isNaN(end) || this.size === 0) {
      this.done = true
      this.end = 0
      this.start = 0
      this.step = 0
    } else if (start > end) {
      this.end = end
      this.start = Number.isFinite(start) ? start : Number.MAX_SAFE_INTEGER
      this.step = -1
    } else {
      this.end = Number.isFinite(end) ? end : Number.MAX_SAFE_INTEGER
      this.start = start
      this.step = 1
    }

    this.state = this.start
  }

  next(): IteratorResult<number, void> {
    if (this.done) {
      return result.done()
    }

    const value = this.state

    if (value === this.end) {
      this.done = true
    } else {
      this.state += this.step
    }

    return result.next(value)
  }

  sizeHint(): number {
    return this.size
  }
}
