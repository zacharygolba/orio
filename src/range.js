// @flow

import { impl } from './iter'
import sizeOf from './utils/size-of'
import type { Iter } from './iter'

type Item = number | string

export default function range(start?: Item = 0, end?: Item = Infinity): Iter<Item> {
  return new Range(start, end)
}

const CharRange = impl(class {
  source: Iterator<number>

  constructor(from: string, to: string) {
    this.source = new NumberRange(from.charCodeAt(0), to.charCodeAt(0))
  }

  next(): IteratorResult<string, void> {
    const result = this.source.next()

    if (result.done) {
      return result
    }

    return {
      done: false,
      value: String.fromCharCode(result.value),
    }
  }

  sizeHint(): number {
    return sizeOf(this.source)
  }
})

const NumberRange = impl(class {
  source: Iterator<number>

  constructor(start: number, end: number) {
    if (start > end) {
      this.source = new DescNumberRange(start, end)
    } else {
      this.source = new AscNumberRange(start, end)
    }
  }

  next(): IteratorResult<number, void> {
    return this.source.next()
  }

  sizeHint(): number {
    return sizeOf(this.source)
  }
})

const AscNumberRange = impl(class {
  end: number
  start: number
  state: number

  constructor(start: number, end: number) {
    this.end = end
    this.start = Number.isFinite(start) ? start : 0
    this.state = this.start
  }

  next(): IteratorResult<number, *> {
    if (this.state > this.end) {
      return {
        done: true,
        value: undefined,
      }
    }

    return {
      done: false,
      value: this.state++,
    }
  }

  sizeHint(): number {
    return Math.abs(this.end - this.start) + 1
  }
})

const DescNumberRange = impl(class {
  end: number
  start: number
  state: number

  constructor(start: number, end: number) {
    this.end = end
    this.start = Number.isFinite(start) ? start : Number.MAX_SAFE_INTEGER
    this.state = this.start
  }

  next(): IteratorResult<number, void> {
    if (this.state < this.end) {
      return {
        done: true,
        value: undefined,
      }
    }

    return {
      done: false,
      value: this.state--,
    }
  }

  sizeHint(): number {
    return Math.abs(this.start - this.end) + 1
  }
})

const Range = impl(class {
  source: Iterator<Item>

  constructor(start: Item, end: Item) {
    if (typeof start === 'string' && typeof end === 'string') {
      this.source = new CharRange(start, end)
    } else {
      this.source = new NumberRange(Number(start), Number(end))
    }
  }

  next(): IteratorResult<Item, void> {
    return this.source.next()
  }

  sizeHint(): number {
    return sizeOf(this.source)
  }
})
