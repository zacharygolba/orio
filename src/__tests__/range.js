// @flow

import range from '../range'

test('range()', () => {
  expect(range(Infinity)).toMatchSnapshot()
  expect(range(Infinity, 0)).toMatchSnapshot()
})

describe('Range<number> ASC', () => {
  let iter

  beforeEach(() => {
    iter = range(1, 3)
  })

  test('#next()', () => {
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
  })

  test('#sizeHint()', () => {
    expect(iter.sizeHint()).toEqual(3)
  })
})

describe('Range<number> DESC', () => {
  let iter

  beforeEach(() => {
    iter = range(3, 1)
  })

  test('#next()', () => {
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
  })

  test('#sizeHint()', () => {
    expect(iter.sizeHint()).toEqual(3)
  })
})

describe('Range<string> ASC', () => {
  let iter

  beforeEach(() => {
    iter = range('a', 'c')
  })

  test('#next()', () => {
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
  })

  test('#sizeHint()', () => {
    expect(iter.sizeHint()).toEqual(3)
  })
})

describe('Range<string> DESC', () => {
  let iter

  beforeEach(() => {
    iter = range('c', 'a')
  })

  test('#next()', () => {
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
    expect(iter.next()).toMatchSnapshot()
  })

  test('#sizeHint()', () => {
    expect(iter.sizeHint()).toEqual(3)
  })
})
