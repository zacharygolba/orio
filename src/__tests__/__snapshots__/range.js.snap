// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range<number> ASC #next() 1`] = `
Object {
  "done": false,
  "value": 1,
}
`;

exports[`Range<number> ASC #next() 2`] = `
Object {
  "done": false,
  "value": 2,
}
`;

exports[`Range<number> ASC #next() 3`] = `
Object {
  "done": false,
  "value": 3,
}
`;

exports[`Range<number> ASC #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Range<number> DESC #next() 1`] = `
Object {
  "done": false,
  "value": 3,
}
`;

exports[`Range<number> DESC #next() 2`] = `
Object {
  "done": false,
  "value": 2,
}
`;

exports[`Range<number> DESC #next() 3`] = `
Object {
  "done": false,
  "value": 1,
}
`;

exports[`Range<number> DESC #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Range<string> ASC #next() 1`] = `
Object {
  "done": false,
  "value": "a",
}
`;

exports[`Range<string> ASC #next() 2`] = `
Object {
  "done": false,
  "value": "b",
}
`;

exports[`Range<string> ASC #next() 3`] = `
Object {
  "done": false,
  "value": "c",
}
`;

exports[`Range<string> ASC #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`Range<string> DESC #next() 1`] = `
Object {
  "done": false,
  "value": "c",
}
`;

exports[`Range<string> DESC #next() 2`] = `
Object {
  "done": false,
  "value": "b",
}
`;

exports[`Range<string> DESC #next() 3`] = `
Object {
  "done": false,
  "value": "a",
}
`;

exports[`Range<string> DESC #next() 4`] = `
Object {
  "done": true,
  "value": undefined,
}
`;

exports[`range() 1`] = `
 {
  "source":  {
    "done": true,
    "end": Infinity,
    "start": Infinity,
    "state": Infinity,
    "step": 0,
  },
}
`;

exports[`range() 2`] = `
 {
  "source":  {
    "done": false,
    "end": 0,
    "start": 9007199254740991,
    "state": 9007199254740991,
    "step": -1,
  },
}
`;
